prometheus/src/main/java/com/uber/m3/tally/prometheus/PrometheusReporter.java

```diff
@@ -0,0 +1,484 @@
+                .observe(interval.getSeconds());
+    }
+
+    private void reportTimerHistogram(String name, Map<String, String> tags, Duration interval) {
+        Map<String, String> ttags = tags;
+        if (tags == null) {
+            ttags = Collections.emptyMap();
+        }
+        String collectorName = canonicalMetricId(name, ttags.keySet());
+        if (!registeredHistograms.containsKey(collectorName)) {
+            Histogram histogram = Histogram.build()
+                    .name(name)
+                    .help(String.format("%s histogram", name))
+                    .buckets(defaultBuckets)
+                    .labelNames(collectionToStringArray(ttags.keySet()))
+                    .register(registry);
+            registeredHistograms.put(collectorName, histogram);
+        }
+        registeredHistograms.get(collectorName)
+                .labels(collectionToStringArray(ttags.values()))
+                .observe(interval.getSeconds());
+    }
+
+    /**
+     * Generates a canonical MetricID for a given name+label keys, not values.
+     *
+     * @param name    metric name.
+     * @param tagKeys label keys.
+     * @return canonical metric ID in a form foo+tag1=1,tag2=1.
+     */
+    static String canonicalMetricId(String name, Set<String> tagKeys) {
+        if (name == null) {
+            throw new IllegalArgumentException("metric name cannot be null");
+        }
+        Map<String, String> tags;
+        if (tagKeys == null) {
+            tags = Collections.emptyMap();
+        } else {
+            tags = new HashMap<>(tagKeys.size());
+            for (String key : tagKeys) {
+                tags.put(key, METRIC_ID_KEY_VALUE);
+            }
+        }
+        return keyForPrefixedStringMaps(name, Collections.singletonList(tags));
+    }
+
+    /**
+     * Generates a unique key for a prefix and a list of maps containing tags.
+     * <p>
+     * If a key occurs in multiple maps, keys on the right take precedence.
+     *
+     * @param prefix prefix for unique key.
+     * @param tags   list of maps containing tags.
+     * @return unique key in a format prefix+key1=value1,key2=value2
+     * @throws IllegalArgumentException when {@code prefix} is null.
+     */
+    private static String keyForPrefixedStringMaps(String prefix, List<Map<String, String>> tags) {
+        if (prefix == null) {
+            throw new IllegalArgumentException("prefix cannot be null");
+        }
+        StringBuilder sb = new StringBuilder();
+        sb.append(prefix).append(PREFIX_SPLITTER);
+        if (tags != null) {
+            Set<String> keys = new HashSet<>();
+            for (Map<String, String> curTags : tags) {
+                keys.addAll(curTags.keySet());
+            }
+            String tagsString = keys.stream()
+                    .sorted()
+                    .map(key -> {
+                        for (int i = tags.size() - 1; i >= 0; i--) {
```

Comment:
Hmm so iiuc this code is basically:

for each key across all maps:
  - find the value for that key (rightmost map takes precedence)
  - stringify the KVP

Would it be simpler (and mildly more efficient) for us to just handle the map merging up front and just go over the kvps once in order by key?

Reply:
removed the multi map support.